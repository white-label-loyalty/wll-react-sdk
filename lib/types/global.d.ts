// Global type definitions for React Native environment

declare const __DEV__: boolean;
