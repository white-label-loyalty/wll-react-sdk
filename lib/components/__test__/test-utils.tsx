import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
