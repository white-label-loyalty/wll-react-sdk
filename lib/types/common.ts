import { ImageProps, ViewStyle } from 'react-native';

export type ImagePropsNoSource = Omit<ImageProps, 'source'>;

export type FlexDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type Justify =
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around'
  | 'evenly';

export type Align = 'start' | 'end' | 'center' | 'stretch' | 'baseline';

export type LayoutProps = {
  children: React.ReactNode;
  justify?: Justify;
  align?: Align;
  direction?: FlexDirection;
  style?: ViewStyle;
};

export const justifyMap: Record<Justify, ViewStyle['justifyContent']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  between: 'space-between',
  around: 'space-around',
  evenly: 'space-evenly',
} as const;

export const alignMap: Record<Align, ViewStyle['alignItems']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;
