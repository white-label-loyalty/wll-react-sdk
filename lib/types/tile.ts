export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
