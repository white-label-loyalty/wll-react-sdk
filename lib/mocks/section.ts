import { SectionType, TSection } from '../types/section';
import { Tile, TileType } from '../types/tile';
import { createTileMock } from './tiles';

export type SectionMockConfig = Partial<TSection> & {
  tiles?: Tile[];
};

const defaultConfig: TSection = {
  details: [],
  defaultLocale: 'en',
  name: 'Elite Member Benefits',
  type: SectionType.Grid,
  active: true,
  id: 'a35399a7-580a-4493-afff-24ef4facf3f4',
  createdAt: '2024-11-19T15:50:37.701Z',
  updatedAt: '2024-12-06T11:18:16.842Z',
  priority: 2,
  tiles: [createTileMock(TileType.Content)],
  title: 'Elite Member Benefits',
  locale: 'en',
  pointsSuffix: 'miles',
};

export const createSectionMock = (config?: SectionMockConfig): TSection => {
  const { tiles, ...sectionConfig } = config || {};
  
  // If tiles are provided in config, use those. Otherwise use default tiles
  const mergedTiles = tiles || defaultConfig.tiles;

  return {
    ...defaultConfig,
    ...sectionConfig,
    tiles: mergedTiles,
  };
};

// Helper function to create a section with specific tile types
export const createSectionWithTiles = (tiles: Tile[]): TSection => {
  return createSectionMock({ tiles });
};
