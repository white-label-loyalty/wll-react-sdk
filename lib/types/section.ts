import { Tile } from './tile';

export enum SectionType {
  Grid = 'GRID',
  Banner = 'BANNER',
}

export type TSection = {
  id: string;
  type: SectionType;
  createdAt: string;
  updatedAt: string;
  title?: string;
  description?: string;
  tiles: Tile[];
  pointsMultiplier?: number;
  pointsPrefix?: string | null;
  pointsSuffix?: string | null;
  priority: number;
  active: boolean;
};
