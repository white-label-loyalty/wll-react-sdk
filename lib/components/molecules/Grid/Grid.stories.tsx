import { Meta, StoryFn } from '@storybook/react';
import * as React from 'react';
import { View } from 'react-native';
import { SectionType } from '../../../types/section';
import {
  ProgressType,
  TierTileType,
  TileHeight,
  TileType,
} from '../../../types/tile';
import Grid from './index';

export default {
  title: 'components/molecules/Grid',
  component: Grid,
  decorators: [
    (Story) => (
      <View
        style={{
          width: 1080,
          height: '100%',
          alignContent: 'center',
          justifyContent: 'center',
        }}
      >
        <Story />
      </View>
    ),
  ],
} as Meta;

const Template: StoryFn<typeof Grid> = (args) => <Grid {...args} />;

export const Default = Template.bind({});
Default.args = {
  section: {
    name: 'Section2',
    type: SectionType.Grid,
    active: true,
    pointsMultiplier: 1,
    pointsPrefix: null,
    pointsSuffix: null,
    id: '84739810-37f5-4621-8433-8e0b61e09a5e',
    createdAt: '2024-10-15T13:36:00.204Z',
    updatedAt: '2024-10-15T13:36:00.204Z',
    title: undefined,
    description: undefined,
    priority: 1,
    tiles: [
      {
        tileHeight: TileHeight.Half,
        active: true,
        type: TileType.Points,
        configuration: {
          title: 'Points balance',
          artworkUrl: 'https://i.ibb.co/s2zBvWD/gold.png',
          points: 300,
        },
        id: 'fda02f0c-7aee-4016-8ac7-f0ec6c168106',
        createdAt: '2024-10-15T13:28:49.613Z',
        updatedAt: '2024-10-15T13:28:49.613Z',
        priority: 3,
      },
      {
        tileHeight: TileHeight.Half,
        active: true,
        type: TileType.Tier,
        configuration: {
          type: TierTileType.currentTier,
          tierId: '54f0658a-9409-41f7-a508-ca02c6ac23e1',
          progressType: ProgressType.Name,
          tier: {
            id: '11ff96d2-c87b-4c89-bf5c-996604fc4d28',
            name: 'Bronze',
            description: undefined,
            artworkUrl: undefined,
            pointsRequirement: 100,
            priority: 0,
          },
        },
        id: '89255d20-737a-4aa9-af12-9266b4b1dedf',
        createdAt: '2024-10-15T13:24:28.991Z',
        updatedAt: '2024-10-15T13:24:28.991Z',
        priority: 2,
      },
      {
        tileHeight: TileHeight.Half,
        active: true,
        type: TileType.Tier,
        configuration: {
          type: TierTileType.currentTargetSpecific,
          tierId: '95a6405a-36d9-471e-9ccb-75088c6f2ba6',
          progressType: ProgressType.Points,
          tier: {
            id: '11ff96d2-c87b-4c89-bf5c-996604fc4d28',
            name: 'Bronze',
            description: null,
            artworkUrl: null,
            pointsRequirement: 100,
            priority: 0,
          },
          targetTier: {
            id: '95a6405a-36d9-471e-9ccb-75088c6f2ba6',
            name: 'Silver',
            pointsRequirement: 2000,
            attained: false,
            artworkUrl: null,
            earnedPoints: 300,
            priority: 0,
          },
        },
        id: '11ac9b70-f025-4184-8372-7acbf0ac08cd',
        createdAt: '2024-10-15T13:21:40.133Z',
        updatedAt: '2024-10-15T13:21:40.133Z',
        priority: 1,
      },
      {
        tileHeight: TileHeight.Half,
        active: true,
        type: TileType.Content,
        configuration: {
          title: 'Welcome Graeme',
          description:
            'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        },
        id: '95d0cf59-198b-48b2-9f42-931cd28bee6e',
        createdAt: '2024-10-15T11:38:59.522Z',
        updatedAt: '2024-10-15T11:38:59.522Z',
        priority: 0,
      },
    ],
  },
};
