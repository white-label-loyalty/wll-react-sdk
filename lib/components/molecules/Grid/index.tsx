import React, { useCallback, useMemo } from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';
import { IS_WEB, SCREEN_WIDTH } from '../../../constants/device';
import { GRID_GAP } from '../../../constants/grid';
import { useResponsive } from '../../../context/ResponsiveContext';
import { SectionType, TSection } from '../../../types/section';
import { Tile, TileHeight, TileType } from '../../../types/tile';
import { sortByPriority } from '../../../utils/transforms';
import { TileContainer } from '../../atoms';
import { SectionHeader } from '../../molecules';

type GridProps = {
  section: TSection;
};

const Grid = ({ section }: GridProps) => {
  if (section.type !== SectionType.Grid) {
    return null;
  }

  const { isDesktop } = useResponsive();
  const columnsPerRow = isDesktop ? 4 : 2;

  const gridTiles = useMemo(() => {
    return sortByPriority(
      section.tiles.filter((tile) => tile.type !== TileType.Banner)
    );
  }, [section.tiles]);

  const getTileWidth = useCallback((columns: number) => {
    if (IS_WEB) {
      return {
        width: `calc(${100 / columns}% - ${((columns - 1) * GRID_GAP) / columns}px)`,
        marginBottom: GRID_GAP,
        height: 'auto',
        aspectRatio: 1,
      };
    } else {
      const availableWidth = SCREEN_WIDTH - GRID_GAP * 2;
      const gapSpace = (columns - 1) * GRID_GAP;

      const tileWidth = Math.floor((availableWidth - gapSpace) / columns);

      return {
        width: tileWidth,
        marginBottom: GRID_GAP,
        aspectRatio: 1,
      };
    }
  }, []);

  const renderTileContainers = useCallback(() => {
    const tileContainers: JSX.Element[] = [];
    let currentTiles: Tile[] = [];

    const sortedTiles = sortByPriority(gridTiles);

    sortedTiles.forEach((tile, index) => {
      currentTiles.push(tile);
      const nextTile = sortedTiles[index + 1];

      const shouldStartNewContainer = (
        currentTiles: Tile[],
        currentTile: Tile,
        nextTile?: Tile
      ) => {
        if (currentTiles.length === 2) return true;
        if (currentTile.tileHeight === TileHeight.Full) return true;
        if (currentTile.tileHeight === TileHeight.Half) {
          if (!nextTile) return true;
          if (nextTile.tileHeight === TileHeight.Full) return true;
          if (currentTiles.length === 0) return false;
          if (
            currentTiles.length === 1 &&
            currentTiles[0].tileHeight === TileHeight.Half &&
            nextTile.tileHeight === TileHeight.Half
          ) {
            return false;
          }
        }
        return true;
      };

      if (shouldStartNewContainer(currentTiles, tile, nextTile)) {
        tileContainers.push(
          <View
            key={`container-${index}`}
            style={[
              // @ts-ignore Web uses CSS calc strings for responsive layouts, while ViewStyle expects numbers
              getTileWidth(columnsPerRow) as ViewStyle,
            ]}
          >
            <TileContainer tiles={currentTiles} />
          </View>
        );
        currentTiles = [];
      }
    });

    return tileContainers;
  }, [gridTiles, columnsPerRow, getTileWidth]);

  return (
    <View>
      <SectionHeader title={section.title} description={section.description} />
      <View style={styles.grid}>{renderTileContainers()}</View>
    </View>
  );
};

const styles = StyleSheet.create({
  grid: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    width: '100%',
    justifyContent: 'space-between',
  },
});

export default Grid;
