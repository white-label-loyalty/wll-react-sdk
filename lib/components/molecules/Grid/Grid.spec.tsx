import '@testing-library/jest-dom';
import React from 'react';
import { SectionType } from '../../../types/section';
import { TileHeight, TileType } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import Grid from './index';

const mockSection = {
  id: 'test-section',
  name: 'Misc Tiles',
  type: SectionType.Grid,
  active: true,
  createdAt: '2025-01-27T18:31:01Z',
  updatedAt: '2025-01-27T18:31:01Z',
  visibilityCriteria: null,
  priority: 5,
  title: 'Profile',
  description:
    'Welcome to WLL Explorer! Find out more about our exciting loyalty program.',
  defaultLocale: 'en',
  locale: 'en',
  tiles: [
    {
      id: '1',
      type: TileType.Badge,
      active: true,
      tileHeight: TileHeight.Full,
      priority: 2,
      configuration: {
        name: 'Badge Title',
        description: 'Badge Description',
      },
      createdAt: '2025-01-27T18:31:01Z',
      updatedAt: '2025-01-27T18:31:01Z',
    },
    {
      id: '2',
      type: TileType.Banner,
      active: true,
      tileHeight: TileHeight.Full,
      priority: 2,
      configuration: {
        title: 'Banner Title',
        description: 'Banner Description',
      },
      createdAt: '2025-01-27T18:31:01Z',
      updatedAt: '2025-01-27T18:31:01Z',
    },
  ],
};

describe('<Grid />', () => {
  it('matches snapshot', () => {
    const { container } = render(<Grid section={mockSection} />);
    expect(container).toMatchSnapshot();
  });

  it('renders section header and filters banner tiles', () => {
    const { getByText, queryByText } = render(<Grid section={mockSection} />);

    expect(getByText('Profile')).toBeInTheDocument();
    expect(
      getByText(
        'Welcome to WLL Explorer! Find out more about our exciting loyalty program.'
      )
    ).toBeInTheDocument();

    expect(getByText('Badge Title')).toBeInTheDocument();
    expect(queryByText('Banner Title')).toBeNull();
  });

  it('filters out banner tiles from the grid', () => {
    const { queryAllByTestId } = render(<Grid section={mockSection} />);
    const containers = queryAllByTestId('tile-container');

    // Should only have one container since we filter out the banner tiles
    expect(containers).toHaveLength(1);
  });

  it('returns null when type is not grid', () => {
    const { container } = render(
      <Grid section={{ ...mockSection, type: SectionType.Banner }} />
    );
    expect(container).toBeEmptyDOMElement();
  });
});
