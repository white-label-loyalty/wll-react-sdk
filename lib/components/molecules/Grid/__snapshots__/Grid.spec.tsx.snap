// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> matches snapshot 1`] = `
<div>
  <div
    class="css-view-175oi2r"
  >
    <div
      class="css-view-175oi2r r-marginBottom-1ifxtd0"
    >
      <div
        class="css-text-146c3p1 r-fontSize-1ui5ee8 r-marginBottom-5oul0u"
        dir="auto"
        style="font-weight: 700; color: rgb(0, 0, 0);"
      >
        Profile
      </div>
      <div
        class="css-text-146c3p1 r-fontSize-1x35g6"
        dir="auto"
        style="color: rgba(0, 0, 0, 0.8);"
      >
        Welcome to WLL Explorer! Find out more about our exciting loyalty program.
      </div>
    </div>
    <div
      class="css-view-175oi2r r-flexDirection-18u37iz r-flexWrap-1w6e6rj r-justifyContent-1wtj0ep r-width-13qz1uu"
    >
      <div
        class="css-view-175oi2r"
        style="width: calc(50% - 8px); margin-bottom: 16px; height: auto; margin-right: 16px; overflow-x: hidden; overflow-y: hidden;"
      >
        <div
          class="css-view-175oi2r r-aspectRatio-1ujkv8a r-flex-13awgt0 r-flexDirection-eqz5dr r-width-13qz1uu"
          data-testid="tile-container"
        >
          <div
            class="css-view-175oi2r r-flex-13awgt0 r-width-13qz1uu"
          >
            <button
              aria-disabled="true"
              aria-label="Tile"
              class="css-view-175oi2r r-pointerEvents-12vffkv r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
              disabled=""
              role="button"
              style="background-color: rgb(248, 247, 252); opacity: 1; flex-direction: column; justify-content: flex-start; align-items: stretch;"
              tabindex="-1"
              type="button"
            >
              <div
                class="css-view-175oi2r r-alignItems-1habvwh r-flex-13awgt0 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep r-paddingBottom-kzbkwu r-paddingInline-3o4zer"
              >
                <div
                  aria-label="Badge title: Badge Title"
                  class="css-view-175oi2r"
                  data-testid="badge-tile-title"
                >
                  <div
                    class="css-text-146c3p1 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01 r-whiteSpace-3s2u2q r-wordWrap-1iln25a r-marginBottom-5oul0u"
                    dir="auto"
                    style="color: rgb(0, 0, 0); font-size: 24px; font-weight: bold;"
                  >
                    Badge Title
                  </div>
                </div>
                <div
                  aria-label="Badge description: Badge Description"
                  class="css-view-175oi2r"
                  data-testid="badge-tile-description"
                >
                  <div
                    class="css-text-146c3p1 r-WebkitBoxOrient-8akbws r-display-krxsd3 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01"
                    dir="auto"
                    style="color: rgb(51, 51, 51); font-size: 14px;"
                  >
                    Badge Description
                  </div>
                </div>
                <div
                  class="css-view-175oi2r"
                  style="flex: 1;"
                />
                <div
                  aria-label="Badge earned on Invalid Date"
                  class="css-view-175oi2r r-alignItems-1habvwh r-borderRadius-y47klf r-paddingBlock-cnw61z r-paddingInline-is05cd"
                  data-testid="badge-tile-date-earned"
                  style="background-color: rgba(57, 46, 215, 0.2);"
                >
                  <div
                    class="css-text-146c3p1 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01 r-whiteSpace-3s2u2q r-wordWrap-1iln25a"
                    dir="auto"
                    style="font-size: 12px; color: rgb(255, 255, 255);"
                  >
                    undefined Invalid Date
                  </div>
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
